Sources/NIOSSL/CustomPrivateKey.swift

```diff
@@ -0,0 +1,268 @@
+    ///     be produced.
+    func sign(channel: Channel, algorithm: SignatureAlgorithm, data: ByteBuffer) -> EventLoopFuture<ByteBuffer>
+
+    /// Called to perform a decryption operation.
+    ///
+    /// The data being passed to the call should be decrypted using _raw_ RSA public key decryption, without padding.
+    /// This call will always execute on `channel.eventLoop`.
+    ///
+    /// This function should only be implemented for RSA keys, and then only if you support RSA key exchange. If you
+    /// are only using TLS 1.3 and later, this function is entirely unnecessary and it will never be called.
+    ///
+    /// - parameters:
+    ///     - channel: The `Channel` representing the connection for which we are performing the decryption operation.
+    ///     - data: The data to be decrypted.
+    /// - returns: An `EventLoopFuture` that will be fulfilled with a `ByteBuffer` containing the decrypted bytes, if
+    ///     the decryption operation completes, or that will be failed with a relevant `Error` if the decrypted bytes
+    ///     could not be produced.
+    func decrypt(channel: Channel, data: ByteBuffer) -> EventLoopFuture<ByteBuffer>
+}
+
+/// This is a type-erased wrapper that can be used to encapsulate a NIOSSLCustomPrivateKey and provide it with
+/// hashability and equatability.
+///
+/// While generally speaking type-erasure has some nasty performance problems, we only need the type-erasure for
+/// Hashable conformance, which we don't use in any production code: only the tests use it. To that end, we don't
+/// mind too much that we need to do this.
+@usableFromInline
+internal struct AnyNIOSSLCustomPrivateKey: NIOSSLCustomPrivateKey, Hashable {
+    @usableFromInline let _value: NIOSSLCustomPrivateKey
+
+    @usableFromInline let _equalsFunction: (NIOSSLCustomPrivateKey) -> Bool
+
+    @usableFromInline let _hashFunction: (inout Hasher) -> Void
+
+    @inlinable init<CustomKey: NIOSSLCustomPrivateKey & Hashable>(_ key: CustomKey) {
+        self._value = key
+        self._equalsFunction = { ($0 as? CustomKey) == key }
+        self._hashFunction = { $0.combine(key) }
+    }
+
+    @inlinable var signatureAlgorithms: [SignatureAlgorithm] {
+        return self._value.signatureAlgorithms
+    }
+
+    @inlinable func sign(channel: Channel, algorithm: SignatureAlgorithm, data: ByteBuffer) -> EventLoopFuture<ByteBuffer> {
+        return self._value.sign(channel: channel, algorithm: algorithm, data: data)
+    }
+
+    @inlinable func decrypt(channel: Channel, data: ByteBuffer) -> EventLoopFuture<ByteBuffer> {
+        return self._value.decrypt(channel: channel, data: data)
+    }
+
+    @inlinable func hash(into hasher: inout Hasher) {
+        self._hashFunction(&hasher)
+    }
+
+    @inlinable static func ==(lhs: AnyNIOSSLCustomPrivateKey, rhs: AnyNIOSSLCustomPrivateKey) -> Bool {
+        return lhs._equalsFunction(rhs._value)
+    }
+}
```

Comment:
If we only use it in tests, could this live in the test target?

Reply:
Ah, no: the `Hashable` requirement is on the class that holds this. What I mean is that NIOSSL itself doesn't use the conformance, but we added it as API and can't take it away.

Reply:
Is there a benefit to use `@inlinable` if we operate on an existential anyway?

Reply:
Not really, no. For `hash` and `==` it's _moderately_ useful simply to hoist the logic a bit, but realistically it's not. The `@inlinable`s are a holdover from a previous design. I can remove them if you think it would improve things.

Reply:
I think, I'd prefer if we only used `@inlinable`, if we can actually assume a performance win.